src/Analyser/MutatingScope.php

```diff
@@ -3558,6 +3559,27 @@ private function specifyExpressionType(Expr $expr, Type $type, ?Type $nativeType
 				if ($specifiedExpr === null) {
 					continue;
 				}
+
+				if ($specifiedExpr instanceof MethodCall
+					&& $type instanceof TypeWithClassName
+					&& $specifiedExpr->var instanceof Variable
+					&& is_string($specifiedExpr->var->name)
+					&& $specifiedExpr->var->name === $variableName
+					&& $specifiedExpr->name instanceof Identifier
+				) {
+					$methodReflection = $this->reflectionProvider->getClass($type->getClassName())->getMethod($specifiedExpr->name->name, $this);
+					$methodReturnType = ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+
+					$moreSpecificTypeHolders[$specifiedExprString] = VariableTypeHolder::createYes(
```

Comment:
Can you please explain what the bug is and why this change fixes it? It seems too specific and too hardcoded for this specific test case...

Reply:
I tried explaining it here: https://github.com/phpstan/phpstan-src/pull/1396#issuecomment-1255913061

Reply:
it boils down to
- phpstan knows a moreSpecificType for a method call `$constant->getValue()`, based on $constant of type `ConstantScalarType`
- after that $constant get narrowed to `ConstantStringType` via `instanceof`
- phpstans already known moreSpecificType is not changed, and therefore phpstan doesn't realize that `$constant->getValue()` can only return `string` after the `instanceof ConstantStringType` case

Reply:
the fix should unblock
- https://github.com/phpstan/phpstan-src/pull/1616
- https://github.com/phpstan/phpstan-src/pull/1383